### ISSUE TITLE
Serve static assests

### ISSUE THREAD

**Issue by javidhsueh** [davezuko/react-redux-starter-kit#176]
I'd like to know how to serve static assets like png/jpg, and how to structure them in the right folder.

**Comment by johnjerrico** [comment]
Hi I'm having similar issues, any updates ?

**Comment by davezuko** [comment]
Hey, I'm currently traveling so I haven't had much of a chance to sit down and think of a thoughtful response. I'll try to get back to you early next week; sorry for the delay.

**Comment by johnjerrico** [comment]
hi dave, thanks for your reply.. really great starter pack; I'm just lost on how to put the images that my css needs.. been trying for days with no progress

**Comment by jurekbarth** [comment]
@johnjerrico it's not a perfect answer, but you can just put them in your `src` directory for development. For example put your image in `src/assets/image.png` and then it's available under `http://localhost:3000/assets/image.png`. Hope this helps for now :)

**Comment by johnjerrico** [comment]
update .. hmm it turns out that it's because I put 
`sassLoader : {
    includePaths :[paths.src('styles'),paths.project('node_modules') + '/compass-mixins/lib',      paths.project('node_modules') + '/susy/sass']
  }`
sorry for the trouble

**Comment by davezuko** [comment]
Does includePaths default to the current file directory if no files are found in the paths from `includePaths`? I haven't used the starter kit yet to load any non-JS/Sass, so this is definitely something that I need to look into in a bit more detail.

**Comment by davezuko** [comment]
@johnjerrico so is it working for you now?

@javidhsueh So right now the starter kit does not come with loaders for assets such as images, you would have to add those to `~/build/webpack/_base.js`. If you're just wanting the assets to be served, I'd suggest something similar to what @jurekbarth mentioned (putting them all under `/assets`), and that way you could configure your server to treat those as static assets. As far as how to do that with webpack I'm not quite sure as I've never done it, but I will play around with it today now that I finally have some time.

**Comment by patrickheeney** [comment]
I am still in the process of figuring out the best way to do this. I spent about a day yesterday trying to modify the starter-kit to use webpack middleware instead of webpack dev sever as you can then use express and easily mock apis, proxy, serve assets, etc without dealing with multiple ports. If this is something that could be introduced into starter-kit I would happily submit a PR?

**Comment by jurekbarth** [comment]
@patrickheeney would be really cool to move to webpack middleware as it is more flexible, than webpack dev server alone. Do you have a link to your codebase?

**Comment by JuoCode** [comment]
I do this when development:
```
// /bin/webpack-dev-server.js
require('babel/register');

var path = require('path');
var express = require('express');

const chalk     = require('chalk');
const devServer = require('../build/webpack-dev-server');
const config    = require('../config');

const host = config.get('webpack_host');
const port = config.get('webpack_port');

devServer.use(express.static(path.resolve(__dirname, '../public')));

devServer.listen(port, host, function () {
  console.log(chalk.green(
    `webpack-dev-server is now running at ${host}:${port}.`
  ));
});
```

In production environment, I think just move `/public` dir into `/dist` and place them into nginx's `root`

**Comment by patrickheeney** [comment]
@jurekbarth Its not open source yet unfortunately. 

@JuoCode In my case I have a full express server running during dev and just wanted to add webpack too it. However it could probably work that way by having webpack setup the dev server and passing that into my server bootstrap instead of creating a new instance. Not sure if that would have any side affects yet.

**Comment by davezuko** [comment]
@patrickheeney that'd be awesome to move to the middleware since it's a lot more flexible for people wanting to extend the starter kit. I don't want to get too much more into it than just setting up the dev server, but if you'd like to submit a PR I think a lot of people would benefit from that setup.

**Comment by micky2be** [comment]
@patrickheeney I will also need to move to webpack-middleware as it will provide me a base to build prototypes.
Would be great if you could provide something asap.

**Comment by johnjerrico** [comment]
hi @davezuko it's due to my 'susy/sass' folder;  thanks for your help. . looking forward for the middleware, sounds awesome;

**Comment by patrickheeney** [comment]
Just opened #202 with the middleware solution.

**Comment by jurekbarth** [comment]
@patrickheeney works like a charm! Already using and loving it.

**Comment by davezuko** [comment]
@patrickheeney thanks so much for the PR, it's been merged. I'll keep this ticket open for a little while longer in case it helps other users.

**Comment by davezuko** [comment]
I'll be pulling this into the general FAQ / Resources issue (here: https://github.com/davezuko/react-redux-starter-kit/issues/221) so that it can still remain helpful for people but not pollute the issue tracker. Thanks again.

**Comment by christiansakai** [comment]
Whats the conclusion here on how to add static assets?

**Comment by justsayno** [comment]
Yes where would I put the my assets folder and is there any config for me to get it to work? I don't quite understand the solution here.

**Comment by patrickheeney** [comment]
Now that the starter-kit is running a standard express server, it follows the same conventions as they would be done natively with express. By default it looks to be serving the [~dist](https://github.com/davezuko/react-redux-starter-kit/blob/master/server/app.js#L36). You can also learn more here: http://expressjs.com/en/starter/static-files.html.

**Comment by justsayno** [comment]
I'm speaking in terms of doing it in development. I will be using a different server completely for production so that's not an issue. In the app.js file it looks as though it is using the config from `_development.js` ultimately becuase if you just run `npm start` your NODE_ENV will be `development` and you will hit this sction of code (not the dist folder part):

```
if (config.env === 'development') {
  const webpack       = require('webpack');
  const webpackConfig = require('../build/webpack/development_hot');
  const compiler      = webpack(webpackConfig);

  app.use(require('./middleware/webpack-dev')({
    compiler,
    publicPath : webpackConfig.output.publicPath
  }));
  app.use(require('./middleware/webpack-hmr')({ compiler }));
} 
```

So the path is what is set as the `publicPath` which is 

```
compiler_public_path : `http://${config.server_host}:${config.server_port}`
```

I don't quiet understand where I put my image files to make it work for example if I simply want to have a tag like this in one of my templates:

```
<a href='/' className='header-brand'><img src='assets/logo.png'/></a>
```

Where do I put my assets folder?

**Comment by davezuko** [comment]
@justsayno this will be changing in the near future once I implement a `--hot` CLI flag, which will allow you to use the express server for static assets as normal in development mode. I'm off for the rest of the night, but I'll read through all these comments tomorrow and hopefully offer a better response.

**Comment by mtsr** [comment]
Personally I feel assets belong in the folder of the component using them. If you set up your web pack url-loader correctly, it will not duplicate the same file that occurs in multiple components, so caching isn’t broken.

**Comment by justsayno** [comment]
@mtsr I'm pretty new to webpack. Can you tell me if I need to do anything other than the standard setup in this starter? IE:

```
      { test: /\.woff(\?.*)?$/,  loader: "url-loader?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=application/font-woff" },
      { test: /\.woff2(\?.*)?$/, loader: "url-loader?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=application/font-woff2" },
      { test: /\.ttf(\?.*)?$/,   loader: "url-loader?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=application/octet-stream" },
      { test: /\.eot(\?.*)?$/,   loader: "file-loader?prefix=fonts/&name=[path][name].[ext]" },
      { test: /\.svg(\?.*)?$/,   loader: "url-loader?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=image/svg+xml" },
      { test: /\.(png|jpg)$/,    loader: 'url-loader?limit=8192' }
```

**Comment by mtsr** [comment]
@justsayno The final loader in that list will pretty much handle all your default image use cases. Any images you put in components and `<img src={require('some image.png')} />` will be included as data-url if small enough, and repackaged and given a content-hash name and put into the dist folder, when building. Try a `npm run compile` to see the results.

**Comment by justsayno** [comment]
@mtsr Thanks for that. Will this also work in development mode as well? EG if I add an image like you mention there it will show up when using `npm start`

**Comment by justsayno** [comment]
@mtsr thanks. I understand now. As you may have guessed I am new to the Webpack stuff and this workflow is a bit different to say a gulp/grunt type one. I like it a lot but just getting my head around it.

**Comment by davidsteinberger** [comment]
I'm not able to get the starter kit to serve misc. static assets. Has anyone a working example that I could look at?

**Comment by davezuko** [comment]
@mtsr I was talking to @justingreenberg about static assets such as `robots.txt` and `favicon` and what-have-you. He mentioned putting these in something like `~/src/assets` and then using the file-loader so webpack pulls them over...

My only final question with this, and maybe you've incorporated this into your workflow: where are those imported? Are those imported in `CoreLayout` or `app.js` or somewhere else? Since at some point you do have to tell webpack about them.

**Comment by davidsteinberger** [comment]
AFAIK, placing them in the `src`-folder will only make webpack serve them if you require those in some component: `src/assets/image.jpg` -> `<img src={require ('assets/image.jpg')} />`.

But how can I make the starterkit serve assets referenced in *.css files?

**Comment by mtsr** [comment]
Nope, sorry, haven’t tried that yet.

My issue with a static assets folder, though, is that people could end up putting other files in there, that really should be `require()`d where they are used, instead.

**Comment by davezuko** [comment]
@mtsr how do you deal with the aforementioned assets then? Take `robots.txt` for example.

**Comment by mtsr** [comment]
@davidsteinberger `url()`s in css are handled by the url-loader.

@davezuko I don't, yet...

**Comment by justingreenberg** [comment]
@davezuko robots.txt is not part of the bundle so it would just be copied to build target in production

[file-loader](https://github.com/webpack/file-loader) can be used to `import` or `require` and bundle, but can also just copy files as well using a query ie `file?name=[path][name].[ext]?[hash]&context=/src/static`, i am running out for a couple of hours but i can provide a more detailed explanation when i get back to my desk

**Comment by patrickheeney** [comment]
Other starter-kits typically copy the assets from `/src/static` to `dist/assets`. Express then runs the serve static on the `dist/assets`. You generally don't use webpack, hot-reloading, etc as you are dealing with favicons, robots.txt, etc. I assume you link to them in your app like any other file asset, but point to the local path: `assets/robots.txt`.

**Comment by patrickheeney** [comment]
I opened #305 to demonstrate this. To test hit `http://localhost:3000/robots.txt`. If you want to add an icon for example, add it to `static/icon.png` then add this to `index.html`: `<link rel="apple-touch-icon" href="icon.png">`. They are served at root level so if you add a folder like `static/assets/file.txt` then the relative path should be `/assets/file.txt`.

**Comment by davidsteinberger** [comment]
Thanks @patrickheeney!

**Comment by davezuko** [comment]
@justingreenberg as far as:

> file-loader can be used to import or require and bundle, but can also just copy files as well using a query ie file?name=[path][name].[ext]?[hash]&context=/src/static

Totally makes sense, 100% on board. My problem is: _where_ is that require statement that defines the query? If it's inside of `app.js` or something similar then that seems slightly odd.

**Comment by patrickheeney** [comment]
You would likely need https://github.com/ampedandwired/html-webpack-plugin or something similar. Basically your index.html would be a template that you could require your icons, etc. (Not sure about robots.txt? as that would need to just be copied). So in theory you build the index.html, it requires the assets, you also have a secondary process to copy over any extra static assets. Then finally it gets served.

One thing to note I am using this starter kit as a client and a server, so I would want to make sure the final package and all assets can be served by express.

**Comment by patrickheeney** [comment]
Looks like this is already in the project, so maybe a helper function to read from a dir and run this https://github.com/ampedandwired/html-webpack-plugin/blob/master/index.js#L152.

**Comment by justingreenberg** [comment]
@patrickheeney :+1: script paths could also be removed and injected dynamically using `inject: true`

@davezuko i would just point out that you don't need a declarative `require` to included in the build, for example you could bind the file-loader directly to the static directory, ie using regex or the main config file

i've been swamped since sunday, hopefully i can carve out a few mins tonight to dig into this

**Comment by davezuko** [comment]
@justingreenberg I've been trying to get that to work w/o any success, so if you have a chance to submit a PR I'd really appreciate it.

**Comment by justingreenberg** [comment]
@davezuko webpack-dev-middleware *should* handle those files... anyway, the file loader would really just be used for assets outside the dev lifecycle, like robots.txt, sitemap.xml, favicon.ico, etc. other statics like images would be handled by url-loader  this is on my todos i'm just trying to find the time. will try to get this tonight!

**Comment by Mas-Tool** [comment]
I also had a problem serving static files while running in live development.
I ended up editing the server/app.js and removing the "if (config.compiler_enable_hmr)"
so the line below will be execute even on hmr mode.
app.use(express.static(paths.base(config.dir_dist)))

**Comment by jessep** [comment]
@Mas-Tool Thanks, that worked for me.

**Comment by davezuko** [comment]
You can now just use `~/src/static` for these files without the need for an extra folder in the project root. These files will be copied over to the root of `~/dist` upon compilation, and I added a temporary workaround in the express server so that they are also served during live development. https://github.com/davezuko/react-redux-starter-kit/pull/338 for relevant commits.

**Comment by netpoetica** [comment]
If I run compile and then serve that dist directory, the static assets are properly served without any issue. However, the webpack dev server fails to display images in src/static. If I write a .scss file with a rule

```
a {
  background: url(/img/myimg.png);
}
```

There will be no error, like 404 for example, but there will be no image. I imagine this is because the request of coming from this blob%3html:// protocol that the CSS is being served on, and probably that protocol isn't producing any errors.

Anyone ever come to figure out how you're supposed to use images with this boilerplate, have a recommendation, or ideally, a quick guide on adding images? Someone should really put an img in HomeView.scss so there is an example of how this is done right off the bat.

**Comment by davezuko** [comment]
@netpoetica I will try to get an example in and tweak any existing issues. I admit this isn't something I haven't put as much time into as I'd like for the starter kit, so apologies for the trouble.

**Comment by netpoetica** [comment]
@davezuko No need to apologize amigo, this is a very useful and thoughtful project. I have to confess though that working with this starter kit I have spent equally as much time thinking about quitting development to become a farmer as I have thinking about how awesome some of the features are :-)

Between this thread and [this other](https://github.com/davezuko/react-redux-starter-kit/commit/b1bb27addc6f9b519fa8e1ffb3271c961cd0f460) one, I was able to sort out that I can have two directories - one static/img and one assets/img. In my HTML, I use /img/my-img.png in src tags for example, but in my .scss file, the url() value sort of acts like "require" so that the blob protocol properly handles this in the dev server. I think this still leads to an issue where static assets served from dist/ would not be properly located though, but it is workable until I come up with something better.
